Rubberduck.Parsing/ParserRuleContextExtensions.cs

```diff
@@ -254,6 +254,29 @@ public static bool ContainsTokenIndex(this ParserRuleContext context, int tokenI
             return precedingContext != null;
         }
 
+        /// <summary>
+        /// Returns the context containing the token following the context provided it is of the specified generic type.
+        /// </summary>
+        public static bool TryGetFollowingContext<TContext>(this ParserRuleContext context, out TContext precedingContext) where TContext : ParserRuleContext
+        {
+            precedingContext = null;
+            if (context == null)
+            {
+                return false;
+            }
+
+            var followingTokenIndex = context.Stop.TokenIndex + 1;
+            var ancestorContainingPrecedingIndex = context.GetAncestorContainingTokenIndex(followingTokenIndex);
+
+            if (ancestorContainingPrecedingIndex == null)
+            {
+                return false;
+            }
+
+            precedingContext = ancestorContainingPrecedingIndex.GetDescendentContainingTokenIndex<TContext>(followingTokenIndex);
```

Comment:
You seem to be walking up and down the tree here... Could this be simplified to not traverse possibly that many levels?

Reply:
I do not see how. To me, this seems to be a minimal tree traversal to find the appropriate contexts, except for the fact that `GetDescendentContainingTokenIndex` could be implemented more efficiently.

You have to go up the tree first until you find a node that actually contains the token with the index you want. (The current one cannot contain it because it is the following token.) After that you can find the contained subcontexts of the desired type that contain the token.